File: tests/cli/integration/completions.test.ts
/**
 * Integration tests for `c3p completions <shell>` (claude-code-profiles-0zn).
 *
 * The contract: the emitted script must source cleanly in the target shell
 * (bash/zsh always; fish only when the binary is on PATH). Tab-completion
 * behaviour itself is shell-tested in the smoke runs below — we validate
 * that the function/widget gets registered without errors.
 */

import { execSync, spawn } from "node:child_process";
import { promises as fs } from "node:fs";
import * as os from "node:os";
import * as path from "node:path";

import { afterEach, describe, expect, it } from "vitest";

import { ensureBuilt, runCli } from "./spawn.js";

interface ShellResult {
  stdout: string;
  stderr: string;
  exitCode: number;
}

function which(cmd: string): boolean {
  try {
    const r = execSync(`command -v ${cmd}`, { stdio: "pipe" });
    return r.toString().trim().length > 0;
  } catch {
    return false;
  }
}

function runShell(shell: string, scriptBody: string): Promise<ShellResult> {
  return new Promise((resolve, reject) => {
    const child = spawn(shell, ["-c", scriptBody], { stdio: ["ignore", "pipe", "pipe"] });
    let stdout = "";
    let stderr = "";
    child.stdout.on("data", (d) => (stdout += d.toString()));
    child.stderr.on("data", (d) => (stderr += d.toString()));
    const timer = setTimeout(() => {
      child.kill("SIGTERM");
      reject(new Error(`shell timed out`));
    }, 5000);
    child.on("close", (code) => {
      clearTimeout(timer);
      resolve({ stdout, stderr, exitCode: code ?? -1 });
    });
    child.on("error", reject);
  });
}

const tmpFiles: string[] = [];
afterEach(async () => {
  for (const f of tmpFiles.splice(0)) {
    await fs.rm(f, { force: true });
  }
});

async function writeTmp(prefix: string, content: string): Promise<string> {
  const p = path.join(os.tmpdir(), `${prefix}-${Date.now()}-${Math.random().toString(36).slice(2)}.sh`);
  await fs.writeFile(p, content);
  tmpFiles.push(p);
  return p;
}

describe("completions <shell> (claude-code-profiles-0zn)", () => {
  it("emits non-empty script for each supported shell", async () => {
    await ensureBuilt();
    for (const shell of ["bash", "zsh", "fish"]) {
      const r = await runCli({ args: ["completions", shell] });
      expect(r.exitCode, `${shell} exit`).toBe(0);
      expect(r.stdout.length, `${shell} script empty`).toBeGreaterThan(100);
    }
  });

  it("rejects unsupported shell with exit 1", async () => {
    await ensureBuilt();
    const r = await runCli({ args: ["completions", "powershell"] });
    expect(r.exitCode).toBe(1);
    expect(r.stderr).toContain("bash|zsh|fish");
  });

  it("rejects missing shell argument with exit 1", async () => {
    await ensureBuilt();
    const r = await runCli({ args: ["completions"] });
    expect(r.exitCode).toBe(1);
    expect(r.stderr).toContain("requires a shell");
  });

  it("bash script sources cleanly and registers _c3p", async () => {
    if (!which("bash")) return;
    await ensureBuilt();
    const r = await runCli({ args: ["completions", "bash"] });
    expect(r.exitCode).toBe(0);
    const scriptPath = await writeTmp("ccp-bash", r.stdout);
    const result = await runShell("bash", `source ${scriptPath} && type _c3p`);
    expect(result.exitCode, `stderr: ${result.stderr}`).toBe(0);
    expect(result.stdout).toContain("_c3p");
  });

  it("zsh script sources cleanly and registers _c3p", async () => {
    if (!which("zsh")) return;
    await ensureBuilt();
    const r = await runCli({ args: ["completions", "zsh"] });
    expect(r.exitCode).toBe(0);
    const scriptPath = await writeTmp("ccp-zsh", r.stdout);
    // Initialize compinit before sourcing the function definition, then check
    // the function is callable. -u skips owner checks (tmp files have $USER as owner anyway).
    const result = await runShell(
      "zsh",
      `autoload -Uz compinit && compinit -u && source ${scriptPath} && type _c3p`,
    );
    expect(result.exitCode, `stderr: ${result.stderr}`).toBe(0);
    expect(result.stdout).toContain("_c3p");
  });

  it("fish script sources cleanly when fish is available", async () => {
    if (!which("fish")) return;
    await ensureBuilt();
    const r = await runCli({ args: ["completions", "fish"] });
    expect(r.exitCode).toBe(0);
    const scriptPath = await writeTmp("ccp-fish", r.stdout);
    const result = await runShell(
      "fish",
      `source ${scriptPath} && complete -c c3p | head -1`,
    );
    expect(result.exitCode, `stderr: ${result.stderr}`).toBe(0);
  });

  it("--json wraps the script in a payload (machine consumers)", async () => {
    await ensureBuilt();
    const r = await runCli({ args: ["completions", "bash", "--json"] });
    expect(r.exitCode).toBe(0);
    const payload = JSON.parse(r.stdout) as { shell: string; script: string };
    expect(payload.shell).toBe("bash");
    expect(payload.script).toContain("_c3p");
  });
});

File: tests/integration/completions_test.go
package integration_test

import (
	"context"
	"encoding/json"
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"strings"
	"testing"
	"time"

	"github.com/htxryan/c3p/tests/integration/helpers"
)

// TestCompletions_NonEmptyScriptForEachShell — IV translation of TS
// completions.test.ts "emits non-empty script for each supported shell".
// The shell completion verbs return scripts >100 bytes for bash/zsh/fish.
func TestCompletions_NonEmptyScriptForEachShell(t *testing.T) {
	helpers.EnsureBuilt(t)
	for _, shell := range []string{"bash", "zsh", "fish"} {
		t.Run(shell, func(t *testing.T) {
			r := mustRun(t, helpers.SpawnOptions{Args: []string{"completions", shell}})
			if r.ExitCode != 0 {
				t.Fatalf("%s exit: want 0, got %d (stderr=%q)", shell, r.ExitCode, r.Stderr)
			}
			if len(r.Stdout) <= 100 {
				t.Fatalf("%s script too short (%d bytes): %q", shell, len(r.Stdout), r.Stdout)
			}
		})
	}
}

// TestCompletions_RejectsUnsupportedShell — exit 1 + bash|zsh|fish
// guidance in stderr.
func TestCompletions_RejectsUnsupportedShell(t *testing.T) {
	helpers.EnsureBuilt(t)
	r := mustRun(t, helpers.SpawnOptions{Args: []string{"completions", "powershell"}})
	if r.ExitCode != 1 {
		t.Fatalf("powershell: want 1, got %d (stderr=%q)", r.ExitCode, r.Stderr)
	}
	if !strings.Contains(r.Stderr, "bash|zsh|fish") {
		t.Errorf("stderr missing 'bash|zsh|fish' enumeration: %q", r.Stderr)
	}
}

// TestCompletions_RejectsMissingShellArg — exit 1 + 'requires a shell'
// message.
func TestCompletions_RejectsMissingShellArg(t *testing.T) {
	helpers.EnsureBuilt(t)
	r := mustRun(t, helpers.SpawnOptions{Args: []string{"completions"}})
	if r.ExitCode != 1 {
		t.Fatalf("completions (no arg): want 1, got %d (stderr=%q)", r.ExitCode, r.Stderr)
	}
	if !strings.Contains(r.Stderr, "requires a shell") {
		t.Errorf("stderr missing 'requires a shell': %q", r.Stderr)
	}
}

// TestCompletions_BashSourcesCleanly — the emitted bash script must
// source without error and register the _c3p completion function.
func TestCompletions_BashSourcesCleanly(t *testing.T) {
	if !shellAvailable("bash") {
		t.Skip("bash not on PATH")
	}
	helpers.EnsureBuilt(t)
	r := mustRun(t, helpers.SpawnOptions{Args: []string{"completions", "bash"}})
	if r.ExitCode != 0 {
		t.Fatalf("completions bash: want 0, got %d (stderr=%q)", r.ExitCode, r.Stderr)
	}
	scriptPath := writeShellTmp(t, "ccp-bash", r.Stdout)
	// Go bin's bash completion registers _c3p_complete (via
	// `complete -F _c3p_complete c3p`). TS used _c3p; we assert against
	// the actual emitted function name.
	out := runShell(t, "bash", fmt.Sprintf("source %s && type _c3p_complete", scriptPath))
	if out.ExitCode != 0 {
		t.Fatalf("bash source: want 0, got %d (stderr=%q)", out.ExitCode, out.Stderr)
	}
	if !strings.Contains(out.Stdout, "_c3p_complete") {
		t.Errorf("bash registered function missing _c3p_complete: %q", out.Stdout)
	}
}

// TestCompletions_ZshSourcesCleanly — same for zsh, with compinit prep.
func TestCompletions_ZshSourcesCleanly(t *testing.T) {
	if !shellAvailable("zsh") {
		t.Skip("zsh not on PATH")
	}
	helpers.EnsureBuilt(t)
	r := mustRun(t, helpers.SpawnOptions{Args: []string{"completions", "zsh"}})
	if r.ExitCode != 0 {
		t.Fatalf("completions zsh: want 0, got %d (stderr=%q)", r.ExitCode, r.Stderr)
	}
	scriptPath := writeShellTmp(t, "ccp-zsh", r.Stdout)
	out := runShell(t, "zsh", fmt.Sprintf(
		"autoload -Uz compinit && compinit -u && source %s && type _c3p", scriptPath))
	if out.ExitCode != 0 {
		t.Fatalf("zsh source: want 0, got %d (stderr=%q)", out.ExitCode, out.Stderr)
	}
	if !strings.Contains(out.Stdout, "_c3p") {
		t.Errorf("zsh registered widget missing _c3p: %q", out.Stdout)
	}
}

// TestCompletions_FishSourcesCleanly — when fish is available.
func TestCompletions_FishSourcesCleanly(t *testing.T) {
	if !shellAvailable("fish") {
		t.Skip("fish not on PATH")
	}
	helpers.EnsureBuilt(t)
	r := mustRun(t, helpers.SpawnOptions{Args: []string{"completions", "fish"}})
	if r.ExitCode != 0 {
		t.Fatalf("completions fish: want 0, got %d (stderr=%q)", r.ExitCode, r.Stderr)
	}
	scriptPath := writeShellTmp(t, "ccp-fish", r.Stdout)
	out := runShell(t, "fish", fmt.Sprintf(
		"source %s && complete -c c3p | head -1", scriptPath))
	if out.ExitCode != 0 {
		t.Fatalf("fish source: want 0, got %d (stderr=%q)", out.ExitCode, out.Stderr)
	}
}

// TestCompletions_JSONWrapsScript — --json wraps the script in
// {shell, script} for machine consumers.
func TestCompletions_JSONWrapsScript(t *testing.T) {
	helpers.EnsureBuilt(t)
	r := mustRun(t, helpers.SpawnOptions{Args: []string{"completions", "bash", "--json"}})
	if r.ExitCode != 0 {
		t.Fatalf("completions bash --json: want 0, got %d (stderr=%q)", r.ExitCode, r.Stderr)
	}
	var payload struct {
		Shell  string `json:"shell"`
		Script string `json:"script"`
	}
	if err := json.Unmarshal([]byte(strings.TrimSpace(r.Stdout)), &payload); err != nil {
		t.Fatalf("parse JSON: %v (raw=%q)", err, r.Stdout)
	}
	if payload.Shell != "bash" {
		t.Errorf("payload.shell: want 'bash', got %q", payload.Shell)
	}
	// _c3p_complete is the bash function; _c3p substring matches it too,
	// but make the assertion explicit.
	if !strings.Contains(payload.Script, "_c3p_complete") {
		t.Errorf("payload.script missing '_c3p_complete': %q", payload.Script)
	}
}

// runShell shells out to `shell -c <body>` with a 15s timeout, matching
// helpers.RunCli's default. zsh `compinit` on a cold runner regularly
// approaches 5s on its own (rebuilding the function digest), so a tighter
// budget here flakes without signal value. Skips the test on context
// deadline rather than failing — completions are a "shell is sane"
// integration check, not a perf gate.
func runShell(t *testing.T, shell, body string) helpers.SpawnResult {
	t.Helper()
	ctx, cancel := context.WithTimeout(context.Background(), 15*time.Second)
	defer cancel()
	cmd := exec.CommandContext(ctx, shell, "-c", body)
	out, err := cmd.Output()
	res := helpers.SpawnResult{Stdout: string(out)}
	if exitErr, ok := err.(*exec.ExitError); ok {
		res.Stderr = string(exitErr.Stderr)
		res.ExitCode = exitErr.ExitCode()
	} else if err != nil {
		if ctx.Err() == context.DeadlineExceeded {
			t.Skipf("runShell %s timed out after 15s; runner too slow for shell-completion smoke test", shell)
		}
		t.Fatalf("runShell %s: %v", shell, err)
	}
	if cmd.ProcessState != nil && res.ExitCode == 0 {
		res.ExitCode = cmd.ProcessState.ExitCode()
	}
	return res
}

// writeShellTmp writes content to a unique temp file under t.TempDir()
// and returns the path. t.TempDir cleanup removes it at test end.
func writeShellTmp(t *testing.T, prefix, content string) string {
	t.Helper()
	p := filepath.Join(t.TempDir(), fmt.Sprintf("%s-%d.sh", prefix, time.Now().UnixNano()))
	if err := os.WriteFile(p, []byte(content), 0o644); err != nil {
		t.Fatalf("write tmp: %v", err)
	}
	return p
}
